### PR TITLE
Purescript 14

### PR DESCRIPTION
**Pull request by srghma**
Includes: 
- https://github.com/f-o-a-m/purescript-web3/pull/154 
- https://github.com/f-o-a-m/purescript-web3/pull/142

but there is an error

```
$ drmaci && docker run -d -p 8545:8545 -e ACCOUNTS_TO_CREATE=10 foamspace/cliquebait:v1.9.12 && timeout 22 sh -c 'until nc -z $0 $1; do sleep 1; done' localhost 8545 && sleep 6 && npm -s test
........

Filters
........
  ✗ Case [Past, ∞] With Trail:

  Error in forked web3 process (RemoteError "Error: [ethjs-provider-http] Invalid JSON RPC response from provider\n    host: http://localhost:8545\n    response:  \"\"\n    responseURL: undefined\n    status: 0\n    statusText: \n  \n    at invalidResponseError (/home/srghma/projects/purescript-web3/node_modules/ethjs-provider-http/lib/index.js:43:23)\n    at XMLHttpRequest.request.onreadystatechange (/home/srghma/projects/purescript-web3/node_modules/ethjs-provider-http/lib/index.js:88:17)\n    at XMLHttpRequestEventTarget.dispatchEvent (/home/srghma/projects/purescript-web3/node_modules/xhr2/lib/xhr2.js:64:18)\n    at XMLHttpRequest._setReadyState (/home/srghma/projects/purescript-web3/node_modules/xhr2/lib/xhr2.js:353:12)\n    at XMLHttpRequest._onHttpRequestError (/home/srghma/projects/purescript-web3/node_modules/xhr2/lib/xhr2.js:542:12)\n    at ClientRequest.<anonymous> (/home/srghma/projects/purescript-web3/node_modules/xhr2/lib/xhr2.js:413:24)\n    at ClientRequest.emit (node:events:365:28)\n    at Socket.socketErrorListener (node:_http_client:447:9)\n    at Socket.emit (node:events:365:28)\n    at emitErrorNT (node:internal/streams/destroy:193:8)")

```

On v1.9.14

```
  ✗ Case [Past, ∞] No Trail:

  Error in forked web3 process (RemoteError "Error: [ethjs-provider-http] Invalid JSON RPC response from provider\n    host: http://localhost:8545\n    response:  \"\"\n    responseURL: undefined\n    status: 0\n    statusText: \n  \n    at invalidResponseError (/home/srghma/projects/purescript-web3/node_modules/ethjs-provider-http/lib/index.js:43:23)\n    at XMLHttpRequest.request.onreadystatechange (/home/srghma/projects/purescript-web3/node_modules/ethjs-provider-http/lib/index.js:88:17)\n    at XMLHttpRequestEventTarget.dispatchEvent (/home/srghma/projects/purescript-web3/node_modules/xhr2/lib/xhr2.js:64:18)\n    at XMLHttpRequest._setReadyState (/home/srghma/projects/purescript-web3/node_modules/xhr2/lib/xhr2.js:353:12)\n    at XMLHttpRequest._onHttpRequestError (/home/srghma/projects/purescript-web3/node_modules/xhr2/lib/xhr2.js:542:12)\n    at ClientRequest.<anonymous> (/home/srghma/projects/purescript-web3/node_modules/xhr2/lib/xhr2.js:413:24)\n    at ClientRequest.emit (node:events:365:28)\n    at Socket.socketErrorListener (node:_http_client:447:9)\n    at Socket.emit (node:events:365:28)\n    at emitErrorNT (node:internal/streams/destroy:193:8)")
  ✗ Case [Past, ∞] With Trail:

  Error in forked web3 process (RemoteError "Error: [ethjs-provider-http] Invalid JSON RPC response from provider\n    host: http://localhost:8545\n    response:  \"\"\n    responseURL: undefined\n    status: 0\n    statusText: \n  \n    at invalidResponseError (/home/srghma/projects/purescript-web3/node_modules/ethjs-provider-http/lib/index.js:43:23)\n    at XMLHttpRequest.request.onreadystatechange (/home/srghma/projects/purescript-web3/node_modules/ethjs-provider-http/lib/index.js:88:17)\n    at XMLHttpRequestEventTarget.dispatchEvent (/home/srghma/projects/purescript-web3/node_modules/xhr2/lib/xhr2.js:64:18)\n    at XMLHttpRequest._setReadyState (/home/srghma/projects/purescript-web3/node_modules/xhr2/lib/xhr2.js:353:12)\n    at XMLHttpRequest._onHttpRequestError (/home/srghma/projects/purescript-web3/node_modules/xhr2/lib/xhr2.js:542:12)\n    at ClientRequest.<anonymous> (/home/srghma/projects/purescript-web3/node_modules/xhr2/lib/xhr2.js:413:24)\n    at ClientRequest.emit (node:events:365:28)\n    at Socket.socketErrorListener (node:_http_client:447:9)\n    at Socket.emit (node:events:365:28)\n    at emitErrorNT (node:internal/streams/destroy:193:8)")
```

https://pastebin.com/QvhJBTVN

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,7 +8,6 @@ node_modules/
 /.psa*
 *.swp
 *#*
-package-lock.json
```
Comment (martyall) [inline .gitignore:11]: why are we including this? adaik it is not standard practice in purescript libraries.

File: packages.dhall [hunk 2]
```diff
@@ -26,8 +26,8 @@ let additions =
           , "ring-modules"
           , "simple-json"
           ]
-        , repo = "https://github.com/f-o-a-m/purescript-eth-core.git"
-        , version = "v7.0.0"
+        , repo = "https://github.com/srghma/purescript-eth-core.git"
```
Comment (martyall) [inline packages.dhall:29]: don't forget to change back

File: src/Network/Ethereum/Web3/Solidity/Generic.purs [hunk 3]
```diff
@@ -215,12 +215,9 @@ dParser = do
   lookAhead
     $ do
         (ParseState _ (Position p) _) <- get
-        _ <- take (dataOffset * 2 - (p.column - 1))
+        _ <- takeBytes (dataOffset - (p.column - 1))
```
Comment (martyall) [inline src/Network/Ethereum/Web3/Solidity/Generic.purs:218]: I'm actually surprised this still works: If you look at the definition of `takeByte`, the parser position state is incremented once per byte. Yet the definition of `dParser` is using the position relative to the characters 🤔

File: src/Network/Ethereum/Web3/Solidity/Size.purs [hunk 4]
```diff
@@ -225,7 +225,7 @@ class KnownSize (d :: DigitList) where
   sizeVal :: DLProxy d -> Int
 
 instance knownSizeBase :: KnownDigit head => KnownSize (DOne head) where
-  sizeVal p = digitVal (DProxy :: DProxy head)
+  sizeVal _p = digitVal (DProxy :: DProxy head)
```
Comment (martyall) [inline src/Network/Ethereum/Web3/Solidity/Size.purs:228]: The argument isn't even being used, should just be `_`

File: test/web3/Main.purs [hunk 5]
```diff
@@ -34,26 +34,26 @@ main =
     p <- liftEffect $ httpProvider "http://localhost:8545"
     join
       $ runSpecT cfg [ consoleReporter ] do
```
Comment (martyall) [inline test/web3/Main.purs:36]: why are these being commented out?

File: test/web3/Web3Spec/Live/RPCSpec.purs [hunk 6]
```diff
@@ -133,16 +133,16 @@ spec provider =
...
       Tuple tx tx' <-
         assertWeb3 provider do
           tx <- Api.eth_getTransactionByBlockHashAndIndex txReceipt.blockHash zero
           tx' <- Api.eth_getTransactionByBlockNumberAndIndex (BN txReceipt.blockNumber) zero
           pure $ Tuple tx tx'
       tx `shouldEqual` tx'
 
-signatureToByteString :: Sig.Signature -> BS.ByteString
-signatureToByteString (Sig.Signature sig) = Hex.toByteString sig.r <> Hex.toByteString sig.s <> BS.singleton (mkQuotient sig.v)
+-- | signatureToByteString :: Sig.Signature -> BS.ByteString
```
Comment (martyall) [inline test/web3/Web3Spec/Live/RPCSpec.purs:144]: should be deleted if not being used

**Review by martyall — CHANGES_REQUESTED** [review]
Overall looks good. Just a couple of questions before we move forward.

File: test/web3/Main.purs (hunk 5 above)
Comment (srghma) [reply test/web3/Main.purs:36]: sorry, I was testing error on my computer, will clean this up

File: src/Network/Ethereum/Web3/Solidity/Generic.purs (hunk 3 above)
Comment (srghma) [reply src/Network/Ethereum/Web3/Solidity/Generic.purs:218]: let's change

```
takeByte :: forall m. Monad m => ParserT HexString m HexString
takeByte = do
  ParseState input (Position position) _ <- get
  if hexLength input < 2 then
    fail "Unexpected EOF"
  else do
    let
      { after, before } = splitAt 2 (unHex input)

      unsafeMkHex s = unsafePartial $ fromJust $ mkHexString s

      position' = Position $ position { column = position.column + 1 }
    put $ ParseState (unsafeMkHex after) position' true
    pure $ unsafeMkHex before
```

to

```
takeByte :: forall m. Monad m => ParserT HexString m HexString
takeByte = do
  ParseState input (Position position) _ <- get
  if hexLength input < 2 then
    fail "Unexpected EOF"
  else do
    let
      { after, before } = splitAt 2 (unHex input)

      unsafeMkHex s = unsafePartial $ fromJust $ mkHexString s

      position' = Position $ position { column = position.column + 2 } -- !!!!
    put $ ParseState (unsafeMkHex after) position' true
    pure $ unsafeMkHex before
```

?

File: .gitignore (hunk 1 above)
Comment (srghma) [reply .gitignore]: why are we ignoring `package-lock.json` You mean?

**Comment by martyall** [comment]
Please see [this comment](https://github.com/f-o-a-m/purescript-web3-generator/pull/69?notification_referrer_id=NT_kwDOAd1MEbMzMTM5MzM2NDA1OjMxMjgwMTQ1#issuecomment-1047126298)

**Comment by martyall** [comment]
Closing in favor of #154

**Pull request closed without merging**